### DISCUSSION TITLE
Custom created_by column in history.

### DISCUSSION THREAD

**Discussion by benshalev849** [ngoduykhanh/PowerDNS-Admin#1438] [Ideas]
We have an application that requires the created_by column to be customizable by a header sent.

We have a main user/team that contains multiple people and "sub-teams" in them, this means that we created 1 API-key with high permissions for the main team, but when their sub-teams and smaller individuals make changes we cannot see it, this is due to all of them being under the same description.

We cannot create multiple different API-keys due to the team using a central point/system to edit records and access the powerdns admin.

We wanted for a while to allow for better auditing and better readability on what user makes changes.
This will be possible by allowing more customizability to the history, the change will be made so you can specify a specific header that the history created_by segment will be saved by that specific header, instead of a general API-key description.

Created discussion referencing the following pull request:
https://github.com/PowerDNS-Admin/PowerDNS-Admin/pull/1437

**Comment by AzorianMatt** [comment]
@benshalev849 this makes a lot more sense to me now that it has been documented as such. This is definitely not the take away I got from the initial PR review. If you want to go ahead and re-create that PR based on the `dev` branch as required by the new contribution guidelines, I will accept this change as part of the 0.4.1 milestone.

Thanks!

**Comment by raunz** [comment]
@benshalev849 - I understand the issue You have, but the variable naming in code isn't quite spot on: "custom_header_setting". It would make a lot more sense, if the variable is named something like "changed_by_addition" or something.
Also the actual variable in request should be: "requested_by" or smth. Replacing the apikey.description with custom_header_setting removes all traces to the actual apikey being used - my recommendation would be to add data to the apikey.desc. Then we would have full usable history "Apikey.desc John Doe, special dpt".

**Comment by AzorianMatt** [comment]
@benshalev849 @raunz makes a great point which I didn't catch in my review. Will you implement his proposed changes?

**Comment by benshalev849** [comment]
@AzorianMatt @raunz 

I changed completely the apikey because in my mind if both the admin made a custom header option, and the sender sent it the apikey description is redundant since both admin and user "agreed" to replace it and see it as absolute, in addition to adding it to the current description might break things which i wanted to avoid, i can change it if seems necessary but in my mind, the description is redundant when a header that is made to "replace" it is specified.

Regarding the name of the setting ill change it once decided if the header should be added with the description or not.

**Reply by AzorianMatt** [reply]
@benshalev849 it seems you're both making valid points so I'll have to discuss this further with the contribution team to determine where to take this. Admittedly, this particular feature set is not my strong area so I don't really have any input on the matter.
